# OmicSage — Phase 1: QC Pipeline

**Pipeline steps covered:**
1. Dataset preparation — strip processed outputs, keep raw counts only
2. Data ingestion (`ingest.py`)
3. Quality control (`qc.py`)
4. Report generation (`qc_report.py`)

**Datasets:** CITE-seq BMMC · Multiome BMMC · HCC MTX

**Output format:** `run_qc()` now returns a `MuData` object.
- RNA QC metrics live on `mdata["rna"].obs`
- CITE-seq additionally exposes `mdata["adt"]`
- Multiome additionally exposes `mdata["atac"]`
- Modality detection and GEX subsetting are handled internally — no manual splitting required

---

## 0. Setup

In [1]:
import os, sys, warnings
from pathlib import Path

# Navigate to OmicSage root regardless of where the notebook was launched from
root = Path.cwd()
while not (root / 'pipeline').exists() and root != root.parent:
    root = root.parent
os.chdir(root)
print(f'Working directory: {os.getcwd()}')

warnings.filterwarnings('ignore', category=UserWarning)
warnings.filterwarnings('ignore', category=FutureWarning)

import scanpy as sc
import mudata as md
sc.settings.verbosity = 1
print(f'Python  : {sys.version.split()[0]}')
print(f'Scanpy  : {sc.__version__}')
print(f'MuData  : {md.__version__}')

os.makedirs('reports', exist_ok=True)
os.makedirs('data/processed', exist_ok=True)

Working directory: c:\Users\shoko\OneDrive\Desktop\project\OmicSage
Python  : 3.11.15
Scanpy  : 1.11.5
MuData  : 0.3.7


In [2]:
from pipeline.modules.qc.ingest import load_dataset
from pipeline.modules.qc.qc import run_qc
print('✓ Imports OK')

✓ Imports OK


---
## 1. Dataset Preparation

The public GSE194122 files are **processed** — they contain normalized values,
precomputed PCA/UMAP embeddings, and pipeline-derived QC metrics.

We strip all of that out, keeping only:
- `X` = raw integer counts (from `layers['counts']`)
- `var` = gene metadata (`feature_types`, `gene_id`)
- `obs` = biological metadata only (`cell_type`, donor info, batch)

Ground truth columns (`cell_type`, `GEX_pct_counts_mt`) are kept for validation.

> **Run once only.** Skipped automatically if `_raw_only.h5ad` files already exist.

In [3]:
CITE_RAW     = Path('data/benchmark/GSE194122_cite_raw_only.h5ad')
MULTIOME_RAW = Path('data/benchmark/GSE194122_multiome_raw_only.h5ad')

print(f'CITE raw exists     : {CITE_RAW.exists()}')
print(f'Multiome raw exists : {MULTIOME_RAW.exists()}')

CITE raw exists     : True
Multiome raw exists : True


In [4]:
if not CITE_RAW.exists():
    print('Preparing CITE-seq...')
    !python scripts/prepare_datasets.py --dataset cite
else:
    print('✓ CITE raw already exists — skipping')

if not MULTIOME_RAW.exists():
    print('Preparing Multiome (large file, may take a few minutes)...')
    !python scripts/prepare_datasets.py --dataset multiome
else:
    print('✓ Multiome raw already exists — skipping')

✓ CITE raw already exists — skipping
✓ Multiome raw already exists — skipping


In [5]:
# Verify stripped files
import anndata
for path, name in [(CITE_RAW, 'CITE'), (MULTIOME_RAW, 'Multiome')]:
    if path.exists():
        a = anndata.read_h5ad(path, backed='r')
        print(f'--- {name} ---')
        print(f'  {a.n_obs:,} cells x {a.n_vars:,} features')
        print(f'  obsm    : {list(a.obsm.keys())}  <- should be empty')
        print(f'  layers  : {list(a.layers.keys())}  <- should be empty')
        print(f'  obs     : {list(a.obs.columns)}')
        print()
        a.file.close()

--- CITE ---
  90,261 cells x 14,087 features
  obsm    : []  <- should be empty
  layers  : []  <- should be empty
  obs     : ['cell_type', 'GEX_pct_counts_mt', 'batch', 'DonorID', 'Site', 'Samplename', 'DonorNumber', 'DonorAge', 'DonorBMI', 'DonorBloodType', 'DonorRace', 'Ethnicity', 'DonorGender', 'DonorSmoker', 'Modality', 'is_train', 'sample']

--- Multiome ---
  69,249 cells x 129,921 features
  obsm    : []  <- should be empty
  layers  : []  <- should be empty
  obs     : ['cell_type', 'GEX_pct_counts_mt', 'batch', 'DonorID', 'Site', 'Samplename', 'DonorNumber', 'DonorAge', 'DonorBMI', 'DonorBloodType', 'DonorRace', 'Ethnicity', 'DonorGender', 'DonorSmoker', 'Modality', 'ATAC_nCount_peaks', 'ATAC_atac_fragments', 'ATAC_reads_in_peaks_frac', 'ATAC_blacklist_fraction', 'ATAC_nucleosome_signal', 'sample']



---
## 2. GSE194122 — CITE-seq BMMC
90,261 cells × 14,087 features (RNA + ADT).  
Ground truth: `obs['cell_type']`, `obs['GEX_pct_counts_mt']`

`run_qc()` auto-detects the CITE-seq modality, computes all metrics on RNA features
only, filters cells, then returns a MuData with both `rna` and `adt` modalities.

In [12]:
adata_cite = load_dataset(CITE_RAW, sample_name='GSE194122_BMMC_CITE')

INFO  Loading .h5ad: GSE194122_cite_raw_only.h5ad
INFO  adata.X already contains integer counts — no extraction needed.



──────────────────────────────────────────────────
  OmicSage · Dataset Loaded
──────────────────────────────────────────────────
  Sample   : GSE194122_BMMC_CITE
  Format   : h5ad
  Cells    : 90,261
  Features : 14,087
  Raw from : adata.X (already raw)
  X dtype  : float32
  Layers   : none
──────────────────────────────────────────────────



In [13]:
adata_cite

AnnData object with n_obs × n_vars = 90261 × 14087
    obs: 'cell_type', 'GEX_pct_counts_mt', 'batch', 'DonorID', 'Site', 'Samplename', 'DonorNumber', 'DonorAge', 'DonorBMI', 'DonorBloodType', 'DonorRace', 'Ethnicity', 'DonorGender', 'DonorSmoker', 'Modality', 'is_train', 'sample'
    var: 'feature_types', 'gene_id'
    uns: 'omicsage_source'

In [14]:
print(f'Cells    : {adata_cite.n_obs:,}')
print(f'Features : {adata_cite.n_vars:,}')
print(f'X dtype  : {adata_cite.X.dtype}')
if 'feature_types' in adata_cite.var.columns:
    print('Feature types:')
    for ft, n in adata_cite.var['feature_types'].value_counts().items():
        print(f'  {ft:<25} {n:>8,}')
if 'cell_type' in adata_cite.obs.columns:
    print(f'Cell types: {sorted(adata_cite.obs["cell_type"].unique())}')

Cells    : 90,261
Features : 14,087
X dtype  : float32
Feature types:
  GEX                         13,953
  ADT                            134
Cell types: ['B1 B IGKC+', 'B1 B IGKC-', 'CD14+ Mono', 'CD16+ Mono', 'CD4+ T CD314+ CD45RA+', 'CD4+ T activated', 'CD4+ T activated integrinB7+', 'CD4+ T naive', 'CD8+ T CD49f+', 'CD8+ T CD57+ CD45RA+', 'CD8+ T CD57+ CD45RO+', 'CD8+ T CD69+ CD45RA+', 'CD8+ T CD69+ CD45RO+', 'CD8+ T TIGIT+ CD45RA+', 'CD8+ T TIGIT+ CD45RO+', 'CD8+ T naive', 'CD8+ T naive CD127+ CD26- CD101-', 'Erythroblast', 'G/M prog', 'HSC', 'ILC', 'ILC1', 'Lymph prog', 'MAIT', 'MK/E prog', 'NK', 'NK CD158e1+', 'Naive CD20+ B IGKC+', 'Naive CD20+ B IGKC-', 'Normoblast', 'Plasma cell IGKC+', 'Plasma cell IGKC-', 'Plasmablast IGKC+', 'Plasmablast IGKC-', 'Proerythroblast', 'Reticulocyte', 'T prog cycling', 'T reg', 'Transitional B', 'cDC1', 'cDC2', 'dnT', 'gdT CD158b+', 'gdT TCRVD2+', 'pDC']


In [15]:
# run_qc() detects CITE-seq automatically via feature_types.
# GEX subsetting, metric computation, and cell filtering all happen internally.
# Returns MuData with mdata['rna'] and mdata['adt'].
mdata_cite, metrics_cite = run_qc(
    adata_cite,
    min_genes=200, max_genes=2500, max_mt_pct=5.0,
    remove_doublets=True,
    generate_report=True,
    report_path='reports/qc_GSE194122_cite.html',
    sample_name='GSE194122_BMMC_CITE',
)

INFO  Detected modality: cite
INFO  Split: 13953 RNA features, 134 ADT features
INFO  QC start — 90261 cells × 13953 RNA features
INFO  Mitochondrial genes detected: 13
INFO  Ribosomal genes detected: 94
INFO  Hemoglobin genes detected: 11
INFO  Scrublet: 188 / 90261 cells flagged as doublets (0.2%)
INFO  QC complete — kept 21777 / 90261 cells (removed 68484)
INFO  Generating QC report plots...
INFO  MT% ground-truth correlation: 0.9998
INFO  QC report written → reports\qc_GSE194122_cite.html (419.9 KB)



  ✓ QC report → reports\qc_GSE194122_cite.html  (420 KB)



In [16]:
metrics_cite

{'n_cells_input': 90261,
 'n_cells_output': 21777,
 'n_cells_removed': 68484,
 'n_genes_input': 13953,
 'modality': 'cite',
 'n_removed_low_genes': 1124,
 'n_removed_high_genes': 8710,
 'n_removed_high_mt': 64699,
 'n_removed_doublets': 188,
 'thresholds': {'min_genes': 200,
  'max_genes': 2500,
  'max_mt_pct': 5.0,
  'remove_doublets': True},
 'median_genes_per_cell': 1317.0,
 'median_umi_per_cell': 3917.0,
 'median_mt_pct': 6.3421831130981445,
 'median_ribo_pct': 23.660425186157227,
 'median_hb_pct': 0.12322858721017838,
 'n_mt_genes': 13,
 'n_ribo_genes': 94,
 'n_hb_genes': 11}

In [17]:
# Inspect MuData structure
print(f'Detected modality : {metrics_cite["modality"]}')
print(f'MuData keys       : {list(mdata_cite.mod.keys())}')
print(f'RNA  : {mdata_cite["rna"].n_obs:,} cells × {mdata_cite["rna"].n_vars:,} genes')
if 'adt' in mdata_cite.mod:
    print(f'ADT  : {mdata_cite["adt"].n_obs:,} cells × {mdata_cite["adt"].n_vars:,} proteins')

Detected modality : cite
MuData keys       : ['rna', 'adt']
RNA  : 21,777 cells × 13,953 genes
ADT  : 21,777 cells × 134 proteins


In [18]:
m = metrics_cite
pass_rate = 100 * m['n_cells_output'] / m['n_cells_input']
print('─' * 45)
print(f"  {'Cells input':<25} {m['n_cells_input']:>10,}")
print(f"  {'Cells kept':<25} {m['n_cells_output']:>10,}  ({pass_rate:.1f}%)")
print(f"  {'Removed (low genes)':<25} {m['n_removed_low_genes']:>10,}")
print(f"  {'Removed (high genes)':<25} {m['n_removed_high_genes']:>10,}")
print(f"  {'Removed (high MT%)':<25} {m['n_removed_high_mt']:>10,}")
print(f"  {'Removed (doublets)':<25} {m['n_removed_doublets']:>10,}")
print('─' * 45)
print(f"  {'Median genes/cell':<25} {m['median_genes_per_cell']:>10.0f}")
print(f"  {'Median UMI/cell':<25} {m['median_umi_per_cell']:>10.0f}")
print(f"  {'Median MT%':<25} {m['median_mt_pct']:>10.2f}%")

─────────────────────────────────────────────
  Cells input                   90,261
  Cells kept                    21,777  (24.1%)
  Removed (low genes)            1,124
  Removed (high genes)           8,710
  Removed (high MT%)            64,699
  Removed (doublets)               188
─────────────────────────────────────────────
  Median genes/cell               1317
  Median UMI/cell                 3917
  Median MT%                      6.34%


In [13]:
# Validate MT% against ground truth.
# run_qc() already computed MT% on RNA features only — no manual subsetting needed.
# Use permissive thresholds so nearly all cells pass, allowing full comparison.
import numpy as np

mdata_cite_permissive, _ = run_qc(
    adata_cite,
    min_genes=1, max_genes=999_999, max_mt_pct=100.0,
    remove_doublets=False,
    generate_report=False,
)
adata_rna_permissive = mdata_cite_permissive['rna']

our_mt = adata_rna_permissive.obs['pct_counts_mt'].values
gt_mt  = adata_rna_permissive.obs['GEX_pct_counts_mt'].values
valid  = np.isfinite(our_mt) & np.isfinite(gt_mt)
corr   = float(np.corrcoef(our_mt[valid], gt_mt[valid])[0, 1])
mae    = float(np.mean(np.abs(our_mt[valid] - gt_mt[valid])))
print(f'MT% correlation : {corr:.4f}  {"✓ PASS" if corr > 0.99 else "✗ FAIL"} (> 0.99)')
print(f'MT% MAE         : {mae:.4f}%  {"✓ PASS" if mae < 0.5 else "✗ FAIL"} (< 0.5%)')

INFO  Detected modality: cite
INFO  Split: 13953 RNA features, 134 ADT features
INFO  QC start — 90261 cells × 13953 RNA features
INFO  Mitochondrial genes detected: 13
INFO  Scrublet: 188 / 90261 cells flagged as doublets (0.2%)
INFO  QC complete — kept 90261 / 90261 cells (removed 0)


MT% correlation : 0.9998  ✓ PASS (> 0.99)
MT% MAE         : 0.0266%  ✓ PASS (< 0.5%)


In [19]:
mdata_cite

MuData object with n_obs × n_vars = 21777 × 14087
  2 modalities
    rna:	21777 × 13953
      obs:	'cell_type', 'GEX_pct_counts_mt', 'batch', 'DonorID', 'Site', 'Samplename', 'DonorNumber', 'DonorAge', 'DonorBMI', 'DonorBloodType', 'DonorRace', 'Ethnicity', 'DonorGender', 'DonorSmoker', 'Modality', 'is_train', 'sample', 'n_genes_by_counts', 'total_counts', 'total_counts_mt', 'pct_counts_mt', 'total_counts_ribo', 'pct_counts_ribo', 'total_counts_hb', 'pct_counts_hb', 'doublet_score', 'predicted_doublet'
      var:	'feature_types', 'gene_id', 'mt', 'ribo', 'hb', 'n_cells_by_counts', 'mean_counts', 'pct_dropout_by_counts', 'total_counts'
      uns:	'omicsage_source'
    adt:	21777 × 134
      obs:	'cell_type', 'GEX_pct_counts_mt', 'batch', 'DonorID', 'Site', 'Samplename', 'DonorNumber', 'DonorAge', 'DonorBMI', 'DonorBloodType', 'DonorRace', 'Ethnicity', 'DonorGender', 'DonorSmoker', 'Modality', 'is_train', 'sample'
      var:	'feature_types', 'gene_id'
      uns:	'omicsage_source'

In [20]:
# Save outputs
# RNA → input for normalization (Phase 1 continues here)
# ADT → stored separately, input for ADT QC + CLR normalization (future phase)
mdata_cite['rna'].write_h5ad('data/processed/GSE194122_cite_rna_qc.h5ad')
print('✓ Saved: data/processed/GSE194122_cite_rna_qc.h5ad')

if 'adt' in mdata_cite.mod:
    mdata_cite['adt'].write_h5ad('data/processed/GSE194122_cite_adt_qc.h5ad')
    print('✓ Saved: data/processed/GSE194122_cite_adt_qc.h5ad')
    print('  (ADT features preserved, QC-filtered cells only, ready for CLR normalization)')

... storing 'sample' as categorical
... storing 'sample' as categorical


✓ Saved: data/processed/GSE194122_cite_rna_qc.h5ad
✓ Saved: data/processed/GSE194122_cite_adt_qc.h5ad
  (ADT features preserved, QC-filtered cells only, ready for CLR normalization)


---
## 3. GSE194122 — Multiome BMMC (RNA + ATAC)
69,249 cells × 129,921 features (GEX genes + ATAC peaks).  
Ground truth: `obs['cell_type']`, `obs['GEX_pct_counts_mt']`

`run_qc()` auto-detects the Multiome modality and runs RNA QC only.
ATAC peaks are preserved in `mdata["atac"]` with low-quality cells removed.
ATAC-specific QC (TSS enrichment, fragment size distribution) is covered in Phase 4.

In [21]:
adata_multiome = load_dataset(MULTIOME_RAW, sample_name='GSE194122_BMMC_multiome')

INFO  Loading .h5ad: GSE194122_multiome_raw_only.h5ad
INFO  adata.X already contains integer counts — no extraction needed.



──────────────────────────────────────────────────
  OmicSage · Dataset Loaded
──────────────────────────────────────────────────
  Sample   : GSE194122_BMMC_multiome
  Format   : h5ad
  Cells    : 69,249
  Features : 129,921
  Raw from : adata.X (already raw)
  X dtype  : float32
  Layers   : none
──────────────────────────────────────────────────



In [22]:
print(f'Cells    : {adata_multiome.n_obs:,}')
print(f'Features : {adata_multiome.n_vars:,}')
if 'feature_types' in adata_multiome.var.columns:
    print('Feature types:')
    for ft, n in adata_multiome.var['feature_types'].value_counts().items():
        print(f'  {ft:<25} {n:>8,}')

Cells    : 69,249
Features : 129,921
Feature types:
  ATAC                       116,490
  GEX                         13,431


In [23]:
# run_qc() detects Multiome automatically via feature_types.
# Passes the full mixed object — no manual GEX subsetting required.
# Returns MuData with mdata['rna'] (QC metrics + filtered cells)
# and mdata['atac'] (same filtered cells, ready for Phase 4 ATAC QC).
mdata_multiome, metrics_multiome = run_qc(
    adata_multiome,
    min_genes=200, max_genes=2500, max_mt_pct=5.0,
    remove_doublets=True,
    generate_report=True,
    report_path='reports/qc_GSE194122_multiome.html',
    sample_name='GSE194122_BMMC_multiome',
)

INFO  Detected modality: multiome
INFO  Split: 13431 RNA features, 116490 ATAC features
INFO  QC start — 69249 cells × 13431 RNA features
INFO  Mitochondrial genes detected: 13
INFO  Ribosomal genes detected: 95
INFO  Hemoglobin genes detected: 8
INFO  Scrublet: 181 / 69249 cells flagged as doublets (0.3%)
INFO  QC complete — kept 65075 / 69249 cells (removed 4174)
INFO  Generating QC report plots...
INFO  MT% ground-truth correlation: 0.9999
INFO  QC report written → reports\qc_GSE194122_multiome.html (293.5 KB)



  ✓ QC report → reports\qc_GSE194122_multiome.html  (294 KB)



In [24]:
# Inspect MuData structure
print(f'Detected modality : {metrics_multiome["modality"]}')
print(f'MuData keys       : {list(mdata_multiome.mod.keys())}')
print(f'RNA  : {mdata_multiome["rna"].n_obs:,} cells × {mdata_multiome["rna"].n_vars:,} genes')
if 'atac' in mdata_multiome.mod:
    print(f'ATAC : {mdata_multiome["atac"].n_obs:,} cells × {mdata_multiome["atac"].n_vars:,} peaks')
    print('  (same filtered cells as RNA — ATAC QC deferred to Phase 4)')

Detected modality : multiome
MuData keys       : ['rna', 'atac']
RNA  : 65,075 cells × 13,431 genes
ATAC : 65,075 cells × 116,490 peaks
  (same filtered cells as RNA — ATAC QC deferred to Phase 4)


In [25]:
m = metrics_multiome
pass_rate = 100 * m['n_cells_output'] / m['n_cells_input']
print('─' * 45)
print(f"  {'Cells input':<25} {m['n_cells_input']:>10,}")
print(f"  {'Cells kept':<25} {m['n_cells_output']:>10,}  ({pass_rate:.1f}%)")
print(f"  {'Removed (low genes)':<25} {m['n_removed_low_genes']:>10,}")
print(f"  {'Removed (high genes)':<25} {m['n_removed_high_genes']:>10,}")
print(f"  {'Removed (high MT%)':<25} {m['n_removed_high_mt']:>10,}")
print(f"  {'Removed (doublets)':<25} {m['n_removed_doublets']:>10,}")
print('─' * 45)
print(f"  {'Median genes/cell':<25} {m['median_genes_per_cell']:>10.0f}")
print(f"  {'Median UMI/cell':<25} {m['median_umi_per_cell']:>10.0f}")
print(f"  {'Median MT%':<25} {m['median_mt_pct']:>10.2f}%")

─────────────────────────────────────────────
  Cells input                   69,249
  Cells kept                    65,075  (94.0%)
  Removed (low genes)                6
  Removed (high genes)           2,770
  Removed (high MT%)             1,257
  Removed (doublets)               181
─────────────────────────────────────────────
  Median genes/cell               1015
  Median UMI/cell                 1554
  Median MT%                      0.56%


In [26]:
# Save outputs
# RNA → input for normalization (Phase 1 continues here)
# ATAC → preserved with filtered cells, input for Phase 4 ATAC QC
mdata_multiome['rna'].write_h5ad('data/processed/GSE194122_multiome_rna_qc.h5ad')
print('✓ Saved: data/processed/GSE194122_multiome_rna_qc.h5ad')
print('  (RNA features only, QC-filtered cells, ready for normalization)')

if 'atac' in mdata_multiome.mod:
    mdata_multiome['atac'].write_h5ad('data/processed/GSE194122_multiome_atac_qc.h5ad')
    print('✓ Saved: data/processed/GSE194122_multiome_atac_qc.h5ad')
    print('  (ATAC peaks, same filtered cells, ready for Phase 4 ATAC QC)')

... storing 'sample' as categorical
... storing 'sample' as categorical


✓ Saved: data/processed/GSE194122_multiome_rna_qc.h5ad
  (RNA features only, QC-filtered cells, ready for normalization)
✓ Saved: data/processed/GSE194122_multiome_atac_qc.h5ad
  (ATAC peaks, same filtered cells, ready for Phase 4 ATAC QC)


---
## 4. GSE166635 — HCC scRNA-seq (10x MTX)
Wang et al. 2025 HCC dataset — primary Phase 1 milestone benchmark.

Plain RNA input — `run_qc()` returns MuData with `mdata['rna']` only.

In [6]:
adata_hcc = load_dataset('data/test/GSE166635/HCC1', sample_name='GSE166635_HCC1')

INFO  Loading 10x MEX directory: data\test\GSE166635\HCC1



──────────────────────────────────────────────────
  OmicSage · Dataset Loaded
──────────────────────────────────────────────────
  Sample   : GSE166635_HCC1
  Format   : mtx_dir
  Cells    : 16,077
  Features : 33,694
  Raw from : adata.X (already raw)
  X dtype  : float32
  Layers   : none
──────────────────────────────────────────────────



In [7]:
mdata_hcc, metrics_hcc = run_qc(
    adata_hcc,
    min_genes=200, max_genes=2500, max_mt_pct=5.0,
    remove_doublets=True,
    generate_report=True,
    report_path='reports/qc_GSE166635_HCC1.html',
    sample_name='GSE166635_HCC1',
)

INFO  Detected modality: rna
INFO  QC start — 16077 cells × 33694 RNA features
INFO  Mitochondrial genes detected: 13
INFO  Ribosomal genes detected: 103
INFO  Hemoglobin genes detected: 12
INFO  Scrublet: 0 / 16077 cells flagged as doublets (0.0%)
INFO  QC complete — kept 2118 / 16077 cells (removed 13959)
INFO  Generating QC report plots...
INFO  QC report written → reports\qc_GSE166635_HCC1.html (260.7 KB)



  ✓ QC report → reports\qc_GSE166635_HCC1.html  (261 KB)



In [8]:
print(f'Detected modality : {metrics_hcc["modality"]}')
print(f'MuData keys       : {list(mdata_hcc.mod.keys())}')
print(f'RNA  : {mdata_hcc["rna"].n_obs:,} cells × {mdata_hcc["rna"].n_vars:,} genes')

Detected modality : rna
MuData keys       : ['rna']
RNA  : 2,118 cells × 33,694 genes


In [9]:
m = metrics_hcc
pass_rate = 100 * m['n_cells_output'] / m['n_cells_input']
print('─' * 45)
print(f"  {'Cells input':<25} {m['n_cells_input']:>10,}")
print(f"  {'Cells kept':<25} {m['n_cells_output']:>10,}  ({pass_rate:.1f}%)")
print(f"  {'Removed (low genes)':<25} {m['n_removed_low_genes']:>10,}")
print(f"  {'Removed (high genes)':<25} {m['n_removed_high_genes']:>10,}")
print(f"  {'Removed (high MT%)':<25} {m['n_removed_high_mt']:>10,}")
print(f"  {'Removed (doublets)':<25} {m['n_removed_doublets']:>10,}")
print('─' * 45)
print(f"  {'Median genes/cell':<25} {m['median_genes_per_cell']:>10.0f}")
print(f"  {'Median UMI/cell':<25} {m['median_umi_per_cell']:>10.0f}")
print(f"  {'Median MT%':<25} {m['median_mt_pct']:>10.2f}%")

─────────────────────────────────────────────
  Cells input                   16,077
  Cells kept                     2,118  (13.2%)
  Removed (low genes)               82
  Removed (high genes)           6,184
  Removed (high MT%)            13,460
  Removed (doublets)                 0
─────────────────────────────────────────────
  Median genes/cell               2123
  Median UMI/cell                 7071
  Median MT%                      7.62%


In [10]:
mdata_hcc['rna'].write_h5ad('data/processed/GSE166635_HCC1_qc.h5ad')
print('✓ Saved: data/processed/GSE166635_HCC1_qc.h5ad')

... storing 'sample' as categorical


... storing 'feature_types' as categorical


✓ Saved: data/processed/GSE166635_HCC1_qc.h5ad


---
## 5. Summary

In [29]:
print('QC Reports:')
for r in sorted(Path('reports').glob('qc_*.html')):
    print(f'  {r}  ({r.stat().st_size / 1024:.0f} KB)')
print()
print('Processed files (input for normalization):')
for f in sorted(Path('data/processed').glob('*.h5ad')):
    print(f'  {f}  ({f.stat().st_size / 1e6:.1f} MB)')

QC Reports:
  reports\qc_GSE166635_HCC1.html  (261 KB)
  reports\qc_GSE194122_cite.html  (420 KB)
  reports\qc_GSE194122_multiome.html  (294 KB)

Processed files (input for normalization):
  data\processed\GSE166635_HCC1_qc.h5ad  (27.8 MB)
  data\processed\GSE194122_cite_adt_qc.h5ad  (21.7 MB)
  data\processed\GSE194122_cite_rna_qc.h5ad  (203.2 MB)
  data\processed\GSE194122_multiome_atac_qc.h5ad  (1800.7 MB)
  data\processed\GSE194122_multiome_rna_qc.h5ad  (549.2 MB)


---
*OmicSage · Phase 1 · QC Module*

## Phase 1 — Step 2: Normalization

**Input** : `data/processed/GSE194122_cite_rna_qc.h5ad`  (raw counts in `.X`, output of QC step)

**Steps**
1. Save raw counts → `layers['counts']`
2. Normalize per cell to 10 000 counts (CP10K)
3. Log1p transform
4. Select top 2 000 highly variable genes (`flavor='seurat_v3'`)

**Output** : `data/processed/GSE194122_cite_normalized.h5ad`

In [3]:
 
import scanpy as sc
import matplotlib.pyplot as plt
 
from pipeline.modules.qc.normalize import normalize
 
sc.settings.verbosity = 2
sc.settings.set_figure_params(dpi=100, facecolor='white', figsize=(8, 5))
 
print("Imports OK")
print(f"Scanpy version: {sc.__version__}")

Imports OK
Scanpy version: 1.11.5


In [4]:
# ------------------------------------------------------------------
# Load QC-filtered RNA AnnData
# ------------------------------------------------------------------
QC_FILE = "data/processed/GSE194122_cite_rna_qc.h5ad"
 
adata_rna = sc.read_h5ad(QC_FILE)
print(adata_rna)
print(f"\nX dtype      : {adata_rna.X.dtype}")
print(f"Layers       : {list(adata_rna.layers.keys())}")
print(f"Obs columns  : {list(adata_rna.obs.columns[:8])} ...")
 

AnnData object with n_obs × n_vars = 21777 × 13953
    obs: 'cell_type', 'GEX_pct_counts_mt', 'batch', 'DonorID', 'Site', 'Samplename', 'DonorNumber', 'DonorAge', 'DonorBMI', 'DonorBloodType', 'DonorRace', 'Ethnicity', 'DonorGender', 'DonorSmoker', 'Modality', 'is_train', 'sample', 'n_genes_by_counts', 'total_counts', 'total_counts_mt', 'pct_counts_mt', 'total_counts_ribo', 'pct_counts_ribo', 'total_counts_hb', 'pct_counts_hb', 'doublet_score', 'predicted_doublet'
    var: 'feature_types', 'gene_id', 'mt', 'ribo', 'hb', 'n_cells_by_counts', 'mean_counts', 'pct_dropout_by_counts', 'total_counts'
    uns: 'omicsage_source'

X dtype      : float32
Layers       : []
Obs columns  : ['cell_type', 'GEX_pct_counts_mt', 'batch', 'DonorID', 'Site', 'Samplename', 'DonorNumber', 'DonorAge'] ...


In [5]:
# ------------------------------------------------------------------
# Run normalization
# ------------------------------------------------------------------
adata_norm, metrics = normalize(
    adata_rna,
    batch_key="batch",
    target_sum=1e4,
    n_top_genes=2000,
    hvg_flavor="seurat",
    inplace=False,       # do not mutate adata_rna
)
 
print("=== Normalization Metrics ===")
for k, v in metrics.items():
    print(f"  {k:<40}: {v}")

normalizing counts per cell
    finished (0:00:00)
extracting highly variable genes
    finished (0:00:01)
=== Normalization Metrics ===
  n_cells                                 : 21777
  n_genes                                 : 13953
  n_hvg_selected                          : 2000
  target_sum                              : 10000.0
  hvg_flavor                              : seurat
  batch_key                               : batch
  mean_counts_per_cell_after_norm         : 1674.604248046875
  log1p_applied                           : True
  raw_counts_in_layer                     : counts
  normalized_in_layer                     : logcounts


In [6]:
adata_norm

AnnData object with n_obs × n_vars = 21777 × 13953
    obs: 'cell_type', 'GEX_pct_counts_mt', 'batch', 'DonorID', 'Site', 'Samplename', 'DonorNumber', 'DonorAge', 'DonorBMI', 'DonorBloodType', 'DonorRace', 'Ethnicity', 'DonorGender', 'DonorSmoker', 'Modality', 'is_train', 'sample', 'n_genes_by_counts', 'total_counts', 'total_counts_mt', 'pct_counts_mt', 'total_counts_ribo', 'pct_counts_ribo', 'total_counts_hb', 'pct_counts_hb', 'doublet_score', 'predicted_doublet'
    var: 'feature_types', 'gene_id', 'mt', 'ribo', 'hb', 'n_cells_by_counts', 'mean_counts', 'pct_dropout_by_counts', 'total_counts', 'highly_variable', 'means', 'dispersions', 'dispersions_norm', 'highly_variable_nbatches', 'highly_variable_intersection'
    uns: 'omicsage_source', 'log1p', 'hvg', 'omicsage_normalization'
    layers: 'counts', 'logcounts'

In [7]:
# ------------------------------------------------------------------
# Sanity checks
# ------------------------------------------------------------------
import numpy as np
import scipy.sparse as sp
 
# 1. Raw counts preserved?
raw = adata_norm.layers["counts"]
if sp.issparse(raw): raw = raw.toarray()
print(f"layers['counts'] — min={raw.min():.0f}  max={raw.max():.0f}  dtype={raw.dtype}")
assert raw.min() >= 0 and np.isclose(raw % 1, 0).mean() > 0.95, \
    "WARNING: layers['counts'] does not look like integer counts!"
 
# 2. Log1p applied?
x = adata_norm.X
if sp.issparse(x): x = x.toarray()
print(f"X after norm  — min={x.min():.3f}  max={x.max():.3f}  mean={x.mean():.3f}")
 
# 3. Per-cell sums after reversing log1p ≈ target_sum?
row_sums = np.expm1(x).sum(axis=1)
print(f"Per-cell sums (expm1): mean={row_sums.mean():.1f}  std={row_sums.std():.1f}")
 
# 4. HVG count?
n_hvg = adata_norm.var["highly_variable"].sum()
print(f"HVGs selected : {n_hvg}")
 
print("\n✓ All sanity checks passed")

layers['counts'] — min=0  max=35451  dtype=float32
X after norm  — min=0.000  max=9.117  mean=0.120
Per-cell sums (expm1): mean=10000.0  std=0.0
HVGs selected : 2000

✓ All sanity checks passed


In [9]:
from pipeline.modules.qc.normalization_report import run_normalization_report

run_normalization_report(
    adata_norm=adata_norm,
    metrics=metrics,
    report_path="reports/normalization_report.html",
    dataset_name="GSE194122_CITE",
)

Building normalization report for 'GSE194122_CITE' ...
  Rendering HVG scatter ...
  Rendering library size ...
  Rendering top HVGs ...
  Rendering gene detection ...
Report saved → C:\Users\shoko\OneDrive\Desktop\project\OmicSage\reports\normalization_report.html


'C:\\Users\\shoko\\OneDrive\\Desktop\\project\\OmicSage\\reports\\normalization_report.html'

In [10]:
# ------------------------------------------------------------------
# Save normalized AnnData
# ------------------------------------------------------------------
import os
os.makedirs("data/processed", exist_ok=True)
os.makedirs("figures", exist_ok=True)
 
OUT_FILE = "data/processed/GSE194122_cite_normalized.h5ad"
adata_norm.write_h5ad(OUT_FILE)
print(f"Saved → {OUT_FILE}")
print(adata_norm)
 
# Confirm uns provenance record
prov = adata_norm.uns["omicsage_normalization"]
print(f"\nProvenance record:")
for k, v in prov.items():
    print(f"  {k}: {v}")
 

Saved → data/processed/GSE194122_cite_normalized.h5ad
AnnData object with n_obs × n_vars = 21777 × 13953
    obs: 'cell_type', 'GEX_pct_counts_mt', 'batch', 'DonorID', 'Site', 'Samplename', 'DonorNumber', 'DonorAge', 'DonorBMI', 'DonorBloodType', 'DonorRace', 'Ethnicity', 'DonorGender', 'DonorSmoker', 'Modality', 'is_train', 'sample', 'n_genes_by_counts', 'total_counts', 'total_counts_mt', 'pct_counts_mt', 'total_counts_ribo', 'pct_counts_ribo', 'total_counts_hb', 'pct_counts_hb', 'doublet_score', 'predicted_doublet'
    var: 'feature_types', 'gene_id', 'mt', 'ribo', 'hb', 'n_cells_by_counts', 'mean_counts', 'pct_dropout_by_counts', 'total_counts', 'highly_variable', 'means', 'dispersions', 'dispersions_norm', 'highly_variable_nbatches', 'highly_variable_intersection'
    uns: 'omicsage_source', 'log1p', 'hvg', 'omicsage_normalization'
    layers: 'counts', 'logcounts'

Provenance record:
  target_sum: 10000.0
  n_top_genes: 2000
  hvg_flavor: seurat
  batch_key: batch
  min_mean: 0.01

## Summary

| Step | Result |
|------|--------|
| Raw counts preserved | `layers['counts']` ✓ |
| Normalization | CP10K (target_sum=10 000) ✓ |
| Transform | log1p ✓ |
| HVG selection | 2 000 genes, seurat_v3 ✓ |
| Provenance | `uns['omicsage_normalization']` ✓ |
| Output file | `data/processed/GSE194122_cite_normalized.h5ad` ✓ |

**Next step** → PCA + UMAP (`pipeline/modules/qc/reduce.py`)
 




## Phase 1 — Step 3: Dimensionality Reduction

**Input** : `data/processed/GSE194122_cite_normalized.h5ad`  (log1p values in `.X`, HVGs flagged)

**Steps**
1. PCA on HVG subset only (`n_comps=50`)
2. Auto-select number of PCs for neighbor graph (elbow detection via `kneed`)
3. Compute neighbor graph (`n_neighbors=15`)
4. UMAP embedding
5. Validate against ground-truth embeddings (`obsm['GEX_X_umap']`)

**Output** : `data/processed/GSE194122_cite_reduced.h5ad`

In [ ]:
import scanpy as sc
import numpy as np
import matplotlib.pyplot as plt

from pipeline.modules.qc.reduce import reduce

sc.settings.verbosity = 2
sc.settings.set_figure_params(dpi=100, facecolor='white', figsize=(8, 5))

print("Imports OK")
print(f"Scanpy version: {sc.__version__}")

In [ ]:
# ------------------------------------------------------------------
# Load normalized AnnData
# ------------------------------------------------------------------
NORM_FILE = "data/processed/GSE194122_cite_normalized.h5ad"

adata_norm = sc.read_h5ad(NORM_FILE)
print(adata_norm)
print(f"\nX dtype           : {adata_norm.X.dtype}")
print(f"Layers            : {list(adata_norm.layers.keys())}")
print(f"HVGs flagged      : {adata_norm.var['highly_variable'].sum():,}")
print(f"cell_type in obs  : {'cell_type' in adata_norm.obs.columns}")

In [ ]:
# ------------------------------------------------------------------
# Run dimensionality reduction
# n_pcs=None  → auto-select via elbow detection (kneed)
# ------------------------------------------------------------------
adata_reduced, metrics = reduce(
    adata_norm,
    n_comps=50,
    n_pcs=None,              # auto-select
    n_pcs_method="elbow",    # kneed elbow detection, falls back to variance
    n_neighbors=15,
    run_tsne=False,
    inplace=False,           # do not mutate adata_norm
)

print("=== Dimensionality Reduction Metrics ===")
for k, v in metrics.items():
    if k == "variance_explained_per_pc":
        print(f"  {'variance_explained_per_pc':<40}: [{v[0]:.4f}, {v[1]:.4f}, ... ] ({len(v)} values)")
    else:
        print(f"  {k:<40}: {v}")

In [ ]:
# ------------------------------------------------------------------
# Sanity checks
# ------------------------------------------------------------------

# 1. PCA shape
assert "X_pca" in adata_reduced.obsm, "X_pca missing"
assert adata_reduced.obsm["X_pca"].shape == (adata_reduced.n_obs, 50), \
    f"Unexpected X_pca shape: {adata_reduced.obsm['X_pca'].shape}"
print(f"✓ X_pca shape      : {adata_reduced.obsm['X_pca'].shape}")

# 2. UMAP shape
assert "X_umap" in adata_reduced.obsm, "X_umap missing"
assert adata_reduced.obsm["X_umap"].shape == (adata_reduced.n_obs, 2), \
    f"Unexpected X_umap shape: {adata_reduced.obsm['X_umap'].shape}"
print(f"✓ X_umap shape     : {adata_reduced.obsm['X_umap'].shape}")

# 3. Neighbors graph
assert "connectivities" in adata_reduced.obsp, "connectivities missing"
print(f"✓ Neighbor graph   : connectivities {adata_reduced.obsp['connectivities'].shape}")

# 4. Provenance record
assert "omicsage_reduce" in adata_reduced.uns, "uns['omicsage_reduce'] missing"
prov = adata_reduced.uns["omicsage_reduce"]
print(f"✓ PC selection     : method='{prov['pc_selection_method']}', n_pcs_used={prov['n_pcs_used']}")
print(f"✓ Cumulative var   : {prov['cumulative_variance_explained_by_selected_pcs']*100:.1f}%")

print("\n✓ All sanity checks passed")

In [ ]:
# ------------------------------------------------------------------
# Scree plot — inspect PC selection
# ------------------------------------------------------------------
variance_ratio = np.array(metrics["variance_explained_per_pc"])
cumvar = np.cumsum(variance_ratio)
n_pcs_used = metrics["n_pcs_used"]

fig, axes = plt.subplots(1, 2, figsize=(12, 4))

# Per-PC variance
ax = axes[0]
colors = ["#E07B3A" if i < n_pcs_used else "#AAAAAA" for i in range(len(variance_ratio))]
ax.bar(range(1, len(variance_ratio)+1), variance_ratio * 100, color=colors, width=0.8)
ax.axvline(n_pcs_used + 0.5, color="#C0392B", linestyle="--", linewidth=1.5,
           label=f"Selected: {n_pcs_used} PCs ({metrics['pc_selection_method']})")
ax.set_xlabel("Principal Component"); ax.set_ylabel("Variance explained (%)")
ax.set_title("Scree Plot"); ax.legend(frameon=False)
ax.spines[["top","right"]].set_visible(False)

# Cumulative variance
ax = axes[1]
ax.plot(range(1, len(cumvar)+1), cumvar * 100, color="#4C78A8", linewidth=2)
ax.axvline(n_pcs_used, color="#C0392B", linestyle="--", linewidth=1.5,
           label=f"Selected: {n_pcs_used} PCs")
ax.axhline(cumvar[n_pcs_used-1]*100, color="#C0392B", linestyle=":", linewidth=0.8)
ax.annotate(f"{cumvar[n_pcs_used-1]*100:.1f}%",
            xy=(n_pcs_used, cumvar[n_pcs_used-1]*100),
            xytext=(n_pcs_used+3, cumvar[n_pcs_used-1]*100-5),
            fontsize=9, color="#C0392B",
            arrowprops=dict(arrowstyle="->", color="#C0392B"))
ax.set_xlabel("Number of PCs"); ax.set_ylabel("Cumulative variance (%)")
ax.set_title("Cumulative Variance"); ax.legend(frameon=False)
ax.spines[["top","right"]].set_visible(False)

plt.tight_layout()
plt.show()
print(f"Auto-selected n_pcs = {n_pcs_used} "
      f"(explains {cumvar[n_pcs_used-1]*100:.1f}% of variance)")

In [ ]:
# ------------------------------------------------------------------
# UMAP coloured by QC metrics
# Look for unwanted structure driven by technical variation
# ------------------------------------------------------------------
qc_keys = [k for k in ["n_genes_by_counts", "total_counts", "pct_counts_mt", "doublet_score"]
           if k in adata_reduced.obs.columns]

if qc_keys:
    sc.pl.umap(
        adata_reduced,
        color=qc_keys,
        ncols=2,
        cmap="YlOrRd",
        s=3,
        alpha=0.7,
        title=[k.replace("_", " ").title() for k in qc_keys],
        show=True,
    )
else:
    print("No QC columns found in obs — run QC step first")

In [ ]:
# ------------------------------------------------------------------
# Validate against ground-truth UMAP (from the original paper)
# Ground truth: adata_reduced.obsm['GEX_X_umap']
# Our output  : adata_reduced.obsm['X_umap']
#
# We do NOT expect pixel-perfect agreement (different random seeds,
# n_neighbors, normalization choices). Instead we check:
# 1. Cell-type separation is preserved (same broad structure)
# 2. Pearson correlation of pairwise distances > 0.85
# ------------------------------------------------------------------
import scipy.spatial.distance as dist

if "GEX_X_umap" in adata_reduced.obsm:
    gt_umap  = adata_reduced.obsm["GEX_X_umap"]
    our_umap = adata_reduced.obsm["X_umap"]

    # Sample 2000 cells for speed
    rng = np.random.default_rng(42)
    idx = rng.choice(adata_reduced.n_obs, size=min(2000, adata_reduced.n_obs), replace=False)

    gt_dists  = dist.pdist(gt_umap[idx])
    our_dists = dist.pdist(our_umap[idx])
    corr = float(np.corrcoef(gt_dists, our_dists)[0, 1])

    threshold = 0.70   # permissive — UMAP is stochastic
    status = "✓ PASS" if corr > threshold else "✗ FAIL"
    print(f"Pairwise distance correlation vs ground truth : {corr:.4f}  {status} (>{threshold})")
else:
    print("GEX_X_umap not found in obsm — ground truth validation skipped")
    print("(Expected on GSE194122 — check the dataset was loaded correctly)")

In [ ]:
# ------------------------------------------------------------------
# UMAP coloured by cell type (ground truth)
# This is the main biological validation:
# - Each cell type should form a coherent cluster
# - Closely related types should be adjacent
# ------------------------------------------------------------------
if "cell_type" in adata_reduced.obs.columns:
    sc.pl.umap(
        adata_reduced,
        color="cell_type",
        s=3,
        alpha=0.7,
        legend_loc="on data",
        legend_fontsize=7,
        title="UMAP — Ground-truth cell types",
        show=True,
    )

    # Side-by-side: our UMAP vs ground-truth UMAP
    if "GEX_X_umap" in adata_reduced.obsm:
        # Temporarily add GT UMAP as a named obsm key for sc.pl.embedding
        adata_reduced.obsm["X_umap_gt"] = adata_reduced.obsm["GEX_X_umap"]
        fig, axes = plt.subplots(1, 2, figsize=(16, 6))

        sc.pl.embedding(adata_reduced, basis="umap",    color="cell_type",
                        s=3, alpha=0.6, legend_loc="on data", legend_fontsize=6,
                        title="OmicSage UMAP", ax=axes[0], show=False)
        sc.pl.embedding(adata_reduced, basis="umap_gt", color="cell_type",
                        s=3, alpha=0.6, legend_loc="on data", legend_fontsize=6,
                        title="Ground-truth UMAP (paper)", ax=axes[1], show=False)
        plt.tight_layout()
        plt.show()
else:
    print("cell_type not in obs — biological validation skipped")

In [ ]:
# ------------------------------------------------------------------
# Generate HTML report
# ------------------------------------------------------------------
from reports.reduce_report import run_reduce_report

run_reduce_report(
    adata_reduced=adata_reduced,
    metrics=metrics,
    report_path="reports/reduce_report.html",
    dataset_name="GSE194122_CITE",
    batch_key="batch",       # renders an extra UMAP panel coloured by batch
)

In [ ]:
# ------------------------------------------------------------------
# Save reduced AnnData
# ------------------------------------------------------------------
import os
os.makedirs("data/processed", exist_ok=True)

OUT_FILE = "data/processed/GSE194122_cite_reduced.h5ad"
adata_reduced.write_h5ad(OUT_FILE)
print(f"Saved → {OUT_FILE}")
print(adata_reduced)

# Confirm provenance
prov = adata_reduced.uns["omicsage_reduce"]
print(f"\nProvenance record:")
for k, v in prov.items():
    if k == "variance_explained_per_pc":
        print(f"  {k}: [{v[0]:.4f}, {v[1]:.4f}, ...] ({len(v)} values)")
    else:
        print(f"  {k}: {v}")

### Summary

| Step | Result |
|------|--------|
| PCA | 50 components on HVG subset ✓ |
| PC selection | elbow detection (kneed) ✓ |
| Neighbor graph | kNN (n_neighbors=15) ✓ |
| UMAP | 2D embedding ✓ |
| Ground-truth validation | pairwise distance correlation vs paper UMAP ✓ |
| Provenance | `uns['omicsage_reduce']` ✓ |
| Output file | `data/processed/GSE194122_cite_reduced.h5ad` ✓ |

**Next step** → Leiden clustering (`pipeline/modules/qc/cluster.py`)